plugins/org.eclipse.n4js/src/org/eclipse/n4js/scoping/utils/SourceElementExtensions.java

```diff
@@ -0,0 +1,318 @@
+		List<IdentifiableElement> result = new ArrayList<>();
+		if (element instanceof FunctionOrFieldAccessor) {
+			if (!(element instanceof ArrowFunction)) {
+				result.add(((FunctionOrFieldAccessor) element).getLocalArgumentsVariable());
+			}
+		}
+		return result;
+	}
+
+	/**
+	 * Walk through all contents list of the given element. If the entry in the list is a N4ClassDefinition, a
+	 * N4EnumDeclaration, a ExportedVariableDeclaration or a FunctionDefinition its children are walked, too. If the
+	 * entry in the list is a IdentifiableElement this is also collected. If the entry in the list is a before mentioned
+	 * element or a Expression the iterator is pruned as it is clear that below these elements no further visible
+	 * identifiable elements can be found.
+	 */
+	private List<IdentifiableElement> doCollectVisibleIdentifiableElements(VariableEnvironmentElement start,
+			EObject element, boolean includeBlockScopedElements) {
+
+		if (includeBlockScopedElements) {
+			return doCollectVisibleIdentifiableElementsUncached(start, element, includeBlockScopedElements);
+
+		} else {
+			return cache.get(Pair.of("doCollectVisibleIdentifiableElements?includeBlockScopedElements=false", element),
+					element.eResource(), () -> {
+						return doCollectVisibleIdentifiableElementsUncached(start, element, includeBlockScopedElements);
+					});
+		}
+	}
+
+	private List<IdentifiableElement> doCollectVisibleIdentifiableElementsUncached(VariableEnvironmentElement start,
+			EObject element, boolean includeBlockScopedElements) {
+
+		List<IdentifiableElement> result = new ArrayList<>();
+		TreeIterator<EObject> allContents = element.eAllContents();
+		InternalCFEMapper veeSwitch = new InternalCFEMapper(start, includeBlockScopedElements, allContents, result);
+		while (allContents.hasNext()) {
+			EObject next = allContents.next();
+			veeSwitch.doSwitch(next);
+		}
+		return result;
+	}
+
+	private class InternalCFEMapper extends N4JSSwitch<Boolean> {
+		final VariableEnvironmentElement start;
+		final boolean includeBlockScopedElements;
+		final TreeIterator<EObject> allContents;
+		final List<? super IdentifiableElement> addHere;
+
+		InternalCFEMapper(VariableEnvironmentElement start, boolean includeBlockScopedElements,
+				TreeIterator<EObject> allContents, List<? super IdentifiableElement> addHere) {
+
+			this.start = start;
+			this.includeBlockScopedElements = includeBlockScopedElements;
+			this.allContents = allContents;
+			this.addHere = addHere;
+		}
+
+		@Override
+		public Boolean caseN4ClassDeclaration(N4ClassDeclaration feature) {
+			Type polyfilledOrOriginalType = getTypeOrPolyfilledType(feature);
+			N4ClassDeclaration nonNullClassDecl = feature;
+			if (polyfilledOrOriginalType instanceof TClass) {
+				TClass polyfilledOrOriginalTypeCasted = (TClass) polyfilledOrOriginalType;
+				N4ClassDeclaration n4ClassDecl = (N4ClassDeclaration) polyfilledOrOriginalTypeCasted.getAstElement();
```

Comment:
Since going from types model to AST is problematic (most of the time), we should avoid `#getAstElement()` here, if possible. If I haven't overlooked anything, the AST node `n4ClassDecl` is only used in the following to invoke `#getDefinedType()` on it, which will return `polyfilledOrOriginalTypeCasted`. So going via the AST does not seem necessary.

Reply:
I'll try to get rid of it. But actually this code is only a merge of the xtend file with the same name.